src/main/java/net/clgd/ccemux/rendering/javafx/TerminalCanvas.java

```diff
@@ -0,0 +1,103 @@
+package net.clgd.ccemux.rendering.javafx;
+
+import static net.clgd.ccemux.rendering.TerminalFont.*;
+
+import dan200.computercraft.core.terminal.TextBuffer;
+import javafx.beans.binding.DoubleExpression;
+import javafx.beans.property.ReadOnlyDoubleProperty;
+import javafx.scene.canvas.Canvas;
+import javafx.scene.layout.Pane;
+import javafx.scene.paint.Color;
+import lombok.*;
+import net.clgd.ccemux.emulation.EmulatedTerminal;
+import net.clgd.ccemux.rendering.PaletteAdapter;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class TerminalCanvas extends Pane {
+	private final Canvas canvas;
+	private final EmulatedTerminal terminal;
+	private final JFXTerminalFont font;
+	private final PaletteAdapter<Color> paletteAdapter;
+	private final ReadOnlyDoubleProperty termScale;
+
+	private final DoubleExpression margin;
+	private final DoubleExpression charWidth;
+	private final DoubleExpression charHeight;
+	private final DoubleExpression totalWidth;
+	private final DoubleExpression totalHeight;
+
+	public TerminalCanvas(EmulatedTerminal terminal, JFXTerminalFont font, ReadOnlyDoubleProperty termScale) {
+		this.terminal = terminal;
+		this.font = font;
+		this.paletteAdapter = new PaletteAdapter<>(terminal.getPalette(), Color::color);
+		this.termScale = termScale;
+
+		this.margin = termScale.multiply(BASE_MARGIN);
+		this.charWidth = termScale.multiply(BASE_CHAR_WIDTH);
+		this.charHeight = termScale.multiply(BASE_CHAR_HEIGHT);
+		this.totalWidth = margin.multiply(2).add(charWidth.multiply(terminal.getWidth()));
+		this.totalHeight = margin.multiply(2).add(charHeight.multiply(terminal.getHeight()));
+
+		this.prefWidthProperty().bind(totalWidth);
+		this.prefHeightProperty().bind(totalHeight);
+
+		this.canvas = new Canvas(totalWidth.get(), totalHeight.get());
+		canvas.widthProperty().bind(this.widthProperty());
+		canvas.heightProperty().bind(this.heightProperty());
+
+		canvas.widthProperty().addListener(o -> this.redraw());
+		canvas.heightProperty().addListener(o -> this.redraw());
+
+		this.getChildren().add(canvas);
+	}
+
+	@Override
+	public boolean isResizable() {
+		// makes sure that this element is drawn with a rigid size and not
+		// resized by parents
+		return true;
+	}
+
+	public void redraw() {
+		val g = canvas.getGraphicsContext2D();
```

Comment:
There should probably be a lock on `terminal` here, all the other renderers are doing it.